README.md
# Swedish Bank Account for JavaScript

[![Build Status](https://travis-ci.org/brocc-ab/se-bank-account-js.svg?branch=master)](https://travis-ci.org/brocc-ab/se-bank-account-js)
[![Latest Version](https://img.shields.io/github/release/brocc-ab/se-bank-account-js.svg?style=flat-square)](https://github.com/brocc-ab/se-bank-account-js/releases)
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://github.com/brocc-ab/se-bank-account-js)

This package easily validates and returns details about Swedish bank account numbers using JavaScript. The details and validation rules are based on the documentation provided by Bankgirot, which can be found [here](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf).

## Installation

```bash
npm install se-bank-account
```

## Usage

Enter the account number including clearing number in any format (string or integer) to get information about the account number such as the name of the bank, clearing number, and account number. If an account number is invalid it will return `false`.

Get information about a valid account number:

```js
const { account } = require 'se-bank-account';

// Returns an object if valid, otherwise it returns false.
const acc = account('3300-7505092556');

acc.bank // Nordea Personkonto
acc.clearing // 3300
acc.number // 7505092556
```

Using an invalid account number:

```js
const { account } = require 'se-bank-account';

const acc = account('0000-0000000'); // false
```

## Supported Banks

Following Swedish banks are currently supported:

* Aion Bank SA
* Avanza Bank
* BNP Paribas SA
* Bankgirot
* BlueStep Finans
* Citibank
* DNB Bank
* Danske Bank
* Ekobanken
* Erik Penser
* Forex Bank *(discontinued\*)*
* Handelsbanken
* IKANO Bank
* Ica Banken
* JAK Medlemsbank
* Klarna Bank
* Landshypotek
* Lunar Bank A/S
* Länsförsäkringar Bank
* Lån & Spar Bank Sverige
* Marginalen Bank
* MedMera Bank *(discontinued\*)*
* Multitude Bank plc
* Nordax Bank *(NOBA Bank Group\*\*)*
* Nordea
* Nordnet Bank
* Northmill Bank
* Plusgirot
* Resurs Bank
* Riksgälden
* SBAB
* SEB
* Santander Consumer Bank
* Skandiabanken
* Sparbanken Syd
* Svea Bank
* Swedbank
* Swedbank fd. Sparbanken Öresund
* Ålandsbanken Sverige AB

\* Bank is currently not available in Bankgirot's documentation, available in package for backwards compatibility.\
\*\* Package will return Nordax Bank, even though the company name has changed to NOBA Bank Group.

## Contributing

If you find a bug or a bank that might not be supported, please submit an issue on Github directly under [issues](https://github.com/brocc-ab/se-bank-account-js/issues).

Any contributions are welcome!


CHANGELOG.md
### 1.1.0 (2023-11-10)
## Fixes
- [#2](https://github.com/brocc-ab/se-bank-account-js/issues/2): Bank list updated to align with the specification of Bankgirot.
- [#4](https://github.com/brocc-ab/se-bank-account-js/issues/4): Fixed incorrect Nordea account number length validation.

### 1.0.0 (2020-09-25)

- Initial release


test/utils.test.js
const { expect } = require('chai');
const { mod10, mod11 } = require('../lib/utils');

const validMod10ExamplesFromBankgirot = [
  '3316812057492',
  '55555551',
  '9912346',
];

const invalidatedMod10ExamplesFromBankgirot = [
  '3316812057490',
  '55555550',
  '9912340',
];

const validMod11ExamplesFromBankgirot = [
  '1912763608957',
  '241350',
  '324558',
];

const invalidatedMod11ExamplesFromBankgirot = [
  '1912763608950',
  '241351',
  '324550',
];

describe('test mod 10 method', () => {
  it('should be true using examples provided by Bankgirot', () => {
    validMod10ExamplesFromBankgirot.forEach((input) => {
      expect(mod10(input)).to.be.true;
    });
  });

  it('should be false when Bankgirot examples have been modified', () => {
    invalidatedMod10ExamplesFromBankgirot.forEach((input) => {
      expect(mod10(input)).to.be.false;
    });
  });

  it('should be true when valid Swedish ssn', () => {
    expect(mod10('9906130241')).to.be.true;
  });

  it('should be false when invalid Swedish ssn', () => {
    expect(mod10('9906130242')).to.be.false;
  });
});

describe('test mod 11 method', () => {
  it('should be true using examples provided by Bankgirot', () => {
    validMod11ExamplesFromBankgirot.forEach((input) => {
      expect(mod11(input)).to.be.true;
    });
  });

  it('should be false when Bankgirot examples have been modified', () => {
    invalidatedMod11ExamplesFromBankgirot.forEach((input) => {
      expect(mod11(input)).to.be.false;
    });
  });
});

test/account.test.js
const { expect } = require('chai');
const account = require('../lib/account');

const banks = [
  {
    name: 'Svea Bank',
    account: '9660-1000010'
  },
  {
    name: 'Avanza Bank',
    account: '9550-0100000',
  },
  {
    name: 'Aion Bank SA',
    account: '9580-1000100',
  },
  {
    name: 'BlueStep Finans',
    account: '9680-0000000',
  },
  {
    name: 'BNP Paribas SA',
    account: '9470-1000010',
  },
  {
    name: 'Citibank',
    account: '9040-1000100',
  },
  {
    name: 'Danske Bank',
    account: '1200-0000010',
  },
  {
    name: 'DNB Bank',
    account: '9190-1000100',
  },
  {
    name: 'Ekobanken',
    account: '9700-1000010',
  },
  {
    name: 'Erik Penser',
    account: '9590-0000100',
  },
  {
    name: 'Forex Bank',
    account: '9400-0001000',
  },
  {
    name: 'Ica Banken',
    account: '9270-0010000',
  },
  {
    name: 'IKANO Bank',
    account: '9170-0001000',
  },
  {
    name: 'JAK Medlemsbank',
    account: '9670-0000000'
  },
  {
    name: 'Klarna Bank',
    account: '9780-0000100',
  },
  {
    name: 'Landshypotek',
    account: '9390-0000001'
  },
  {
    name: 'Lunar Bank A/S',
    account: '9710-0000000',
  },
  {
    name: 'Lån & Spar Bank Sverige',
    account: '9630-0000001',
  },
  {
    name: 'Länsförsäkringar Bank',
    account: '3400-0001000',
  },
  {
    name: 'Länsförsäkringar Bank',
    account: '9020-0100000',
  },
  {
    name: 'Marginalen Bank',
    account: '9230-1000001',
  },
  {
    name: 'MedMera Bank',
    account: '9650-1000000',
  },
  {
    name: 'Multitude Bank plc',
    account: '9070-0000100',
  },
  {
    name: 'Nordax Bank',
    account: '9640-0010000',
  },
  {
    name: 'Nordea',
    account: '1100-0000001',
  },
  {
    name: 'Nordea',
    account: '1400-0001000',
  },
  {
    name: 'Nordea',
    account: '3000-1001000',
  },
  {
    name: 'Nordea',
    account: '3410-0100000',
  },
  {
    name: 'Nordea',
    account: '4000-1000000',
  },
  {
    name: 'Nordnet Bank',
    account: '9100-0000100',
  },
  {
    name: 'Northmill Bank',
    account: '9750-1000001',
  },
  {
    name: 'Resurs Bank',
    account: '9280-1000000',
  },
  {
    name: 'Riksgälden',
    account: '9880-0001000',
  },
  {
    name: 'Santander Consumer Bank',
    account: '9460-0010000',
  },
  {
    name: 'SBAB',
    account: '9250-0000001',
  },
  {
    name: 'SEB',
    account: '5000-1001000',
  },
  {
    name: 'SEB',
    account: '9120-0010000',
  },
  {
    name: 'SEB',
    account: '9130-1000000',
  },
  {
    name: 'Skandiabanken',
    account: '9150-0000010',
  },
  {
    name: 'Swedbank',
    account: '7000-1001000',
  },
  {
    name: 'Ålandsbanken Sverige AB',
    account: '2300-0000001',
  },
  {
    name: 'Danske Bank',
    account: '9180-1000000008',
  },
  {
    name: 'Handelsbanken',
    account: '6000-100000010',
  },
  {
    name: 'Nordea/Plusgirot',
    account: '9500-1000000008',
  },
  {
    name: 'Nordea Personkonto',
    account: '3300-7505092556',
  },
  {
    name: 'Nordea Personkonto',
    account: '3782-7505092556',
  },
  {
    name: 'Riksgälden',
    account: '9890-1000000008',
  },
  {
    name: 'Sparbanken Syd',
    account: '9570-1000000008',
  },
  {
    name: 'Swedbank',
    account: '80001-1000000008',
  },
  {
    name: 'Swedbank fd. Sparbanken Öresund',
    account: '9300-1000000008',
  },
  {
    name: 'Bankgirot',
    account: '9900-2000008',
  },
];

const invalidAccounts = [
  '0000-0000000',
  '1400-00010001234',
  '2000-10000040',
  '3301-010000099',
  '3410-0100000000000000000',
  '4000-10000003333'
];

describe('test bank account numbers', () => {
  it('should be valid bank account numbers', () => {
    banks.forEach(bank => {
      const acc = account(bank.account);

      // Split clearing and account number.
      const parts = bank.account.split('-');

      expect(acc.bank).to.equal(bank.name);
      expect(acc.clearing).to.equal(parts[0]);
      expect(acc.number).to.equal(parts[1]);
    });
  });

  it('should be invalid bank account numbers', () => {
    invalidAccounts.forEach(accountNumber => {
      const acc = account(accountNumber);

      expect(acc).to.be.false;
    });
  });
});

lib/account.js
const { mod10, mod11 } = require('./utils');

// @see https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf
const banks = [
  {
    name: 'Svea Bank',
    regex: /^966[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Aion Bank SA',
    regex: /^958[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Avanza Bank',
    regex: /^95[5-6][0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'BlueStep Finans',
    regex: /^968[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'BNP Paribas SA',
    regex: /^947[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Citibank',
    regex: /^904[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Danske Bank',
    regex: /^1[2-3][0-9][0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Danske Bank',
    regex: /^24[0-9][0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'DNB Bank',
    regex: /^(919|926)[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Ekobanken',
    regex: /^970[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Erik Penser',
    regex: /^959[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Forex Bank',
    regex: /^94[0-4][0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Ica Banken',
    regex: /^927[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'IKANO Bank',
    regex: /^917[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'JAK Medlemsbank',
    regex: /^967[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Klarna Bank',
    regex: /^978[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Landshypotek',
    regex: /^939[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Lunar Bank A/S',
    regex: /^971[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Lån & Spar Bank Sverige',
    regex: /^963[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Länsförsäkringar Bank',
    regex: /^(340|906)[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Länsförsäkringar Bank',
    regex: /^902[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Marginalen Bank',
    regex: /^923[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'MedMera Bank',
    regex: /^965[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Multitude Bank plc',
    regex: /^907[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Nordax Bank',
    regex: /^964[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Nordea',
    regex: /^11[0-9]{9}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Nordea',
    regex: /^(1[4-9][0-9]{9}$|20[0-9]{9}$)/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Nordea',
    regex: /^(?!3300)(3[0-3][0-9]{9}$)/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Nordea',
    regex: /^(?!3782)(3[4-9][1-9][0-9]{8}$)/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Nordea',
    regex: /^4[0-9]{10}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Nordnet Bank',
    regex: /^910[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Northmill Bank',
    regex: /^975[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Resurs Bank',
    regex: /^928[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Riksgälden',
    regex: /^988[0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Santander Consumer Bank',
    regex: /^946[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'SBAB',
    regex: /^925[0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'SEB',
    regex: /^5[0-9]{10}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'SEB',
    regex: /^912[0-4][0-9]{7}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'SEB',
    regex: /^91[3-4][0-9]{8}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Skandiabanken',
    regex: /^91[5-6][0-9]{8}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Swedbank',
    regex: /^7[0-9]{10}$/,
    type: 1,
    comment: 1,
  },
  {
    name: 'Ålandsbanken Sverige AB',
    regex: /^23[0-9]{9}$/,
    type: 1,
    comment: 2,
  },
  {
    name: 'Danske Bank',
    regex: /^918[0-9]{11}$/,
    type: 2,
    comment: 1,
    mod10: true,
  },
  {
    name: 'Handelsbanken',
    regex: /^6[0-9]{12}$/,
    type: 2,
    comment: 2,
  },
  {
    name: 'Nordea/Plusgirot',
    regex: /^(95[0-4]|996)[0-9]{8,11}$/,
    type: 2,
    comment: 3,
    mod10: true,
  },
  {
    name: 'Nordea Personkonto',
    regex: /^(3300|3782)[0-9]{10}$/,
    type: 2,
    comment: 1,
    mod10: true,
  },
  {
    name: 'Riksgälden',
    regex: /^989[0-9]{11}$/,
    type: 2,
    comment: 1,
    mod10: true,
  },
  {
    name: 'Sparbanken Syd',
    regex: /^957[0-9]{11}$/,
    type: 2,
    comment: 1,
    mod10: true,
  },
  {
    name: 'Swedbank',
    regex: /^8[0-9]{10,14}$/,
    type: 2,
    comment: 3,
    mod10: true,
    clen: 5,
  },
  {
    name: 'Swedbank fd. Sparbanken Öresund',
    regex: /^93[0-4][0-9]{11}$/,
    type: 2,
    comment: 1,
    mod10: true,
  },
  {
    // Note: Bankgirot is not included in their own documentation for Swedish bank account numbers.
    name: 'Bankgirot',
    regex: /^9900[0-9]{7,8}$/,
    mod10: true,
  },
];

/**
 * Validates bank account number against given bank.
 *
 * @param {string} number
 * @param {object} bank
 * @returns {boolean|object}
 */
function validate(number, bank) {
  const clen = bank.clen ? bank.clen : 4;

  const clearing = number.slice(0, clen);

  const accnum = number.slice(clen);

  // Only check type 1 with comment 1 & 2, otherwise full account number excl. clearing is checked.
  const digits = bank.type === 1 && bank.comment === 1
    ? clearing.slice(1) + accnum
    : bank.type === 1 && bank.comment === 2 ? clearing + accnum
      : accnum;

  const valid = bank.mod10 ? mod10(digits) : mod11(digits);

  if (!valid) {
    return false;
  }

  return {
    bank: bank.name,
    clearing: clearing,
    number: accnum,
  };
}

/**
 * Validate Swedish bank account number.
 *
 * @param {string|number} account
 * @returns {boolean|object}
 */
module.exports = function (account) {
  const number = String(account).replace(/\D/g, '');

  for (let i in banks) {
    if (banks[i].regex.test(number)) {
      return validate(number, banks[i]);
    }
  }

  return false;
};